File: 2025/day05/ruby/solution.rb
#!/usr/bin/env ruby

# Read input from file
input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
input = File.read(input_path).strip

# Parse input into sections
ranges_section, ids_section = input.split("\n\n")

# Parse fresh ingredient ranges
ranges = ranges_section.split("\n").map do |line|
  start_val, end_val = line.split('-').map(&:to_i)
  { start: start_val, end: end_val }
end

# Parse available ingredient IDs
available_ids = ids_section.split("\n").map(&:to_i)

# Part 1: Count how many available ingredient IDs are fresh
def part1(ranges, available_ids)
  fresh_count = 0

  available_ids.each do |id|
    # Check if ID falls within any range
    is_fresh = ranges.any? { |range| id >= range[:start] && id <= range[:end] }
    fresh_count += 1 if is_fresh
  end

  fresh_count
end

# Part 2: Count total unique ingredient IDs covered by all ranges
def part2(ranges)
  # Sort ranges by start position
  sorted_ranges = ranges.sort_by { |r| r[:start] }

  # Merge overlapping ranges
  merged = []

  sorted_ranges.each do |range|
    if merged.empty?
      merged << range.dup
    else
      last = merged.last

      # Check if current range overlaps or is adjacent to the last merged range
      if range[:start] <= last[:end] + 1
        # Merge by extending the end if necessary
        last[:end] = [last[:end], range[:end]].max
      else
        # No overlap, add as new range
        merged << range.dup
      end
    end
  end

  # Sum up the size of each merged range
  total_count = 0
  merged.each do |range|
    count = range[:end] - range[:start] + 1
    total_count += count
  end

  total_count
end

puts "Part 1: #{part1(ranges, available_ids)}"
puts "Part 2: #{part2(ranges)}"

File: 2025/day05/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

// Parse input
const [rangesSection, idsSection] = input.split('\n\n');

// Parse fresh ingredient ranges
const ranges = rangesSection.split('\n').map(line => {
  const [start, end] = line.split('-').map(Number);
  return { start, end };
});

// Parse available ingredient IDs
const availableIds = idsSection.split('\n').map(Number);

// Part 1
function part1() {
  let freshCount = 0;

  for (const id of availableIds) {
    // Check if ID falls within any range
    const isFresh = ranges.some(range => id >= range.start && id <= range.end);
    if (isFresh) {
      freshCount++;
    }
  }

  return freshCount;
}

// Part 2
function part2() {
  // Sort ranges by start position
  const sortedRanges = [...ranges].sort((a, b) => a.start - b.start);

  // Merge overlapping ranges
  const mergedRanges = [];
  let current = sortedRanges[0];

  for (let i = 1; i < sortedRanges.length; i++) {
    const next = sortedRanges[i];

    // Check if ranges overlap or are adjacent
    if (next.start <= current.end + 1) {
      // Merge: extend current range to include next
      current = {
        start: current.start,
        end: Math.max(current.end, next.end)
      };
    } else {
      // No overlap: save current and move to next
      mergedRanges.push(current);
      current = next;
    }
  }
  // Don't forget the last range
  mergedRanges.push(current);

  // Count total unique IDs across all merged ranges
  let totalIds = 0;
  for (const range of mergedRanges) {
    totalIds += (range.end - range.start + 1);
  }

  return totalIds;
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());
